### PR TITLE
lobby integration and mock client demo

### PR DESCRIPTION
**Pull request by droidkfx**
This pull implements the lobby logic, including functionality to join, create, get available games etc... I also updated the mock client to demo these changes.

### CODE REVIEW

**Review by C-Merrill — CHANGES_REQUESTED** [review]
Fix merge conflicts and we can get this in

**Comment by mikamerath** [comment]
@katiesweet, can you verify this builds on Linux?

**Comment by C-Merrill** [comment]
@michaelkamerath are we waiting on that linux build test before we merge?

**Comment by mikamerath** [comment]
@C-Merrill, yes, I asked @katiesweet to do this for builds so we don't have that problem again

File: source/Messages/LobbyGame.hpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#define LOBBY_GAME
+
+#include <string>
+#include <vector>
+#include <memory>
+
+#include <boost\serialization\access.hpp>
+
+#include "source\PlayerAPI\Player.hpp"
+
```
Comment (katiesweet) [inline source/Messages/LobbyGame.hpp:11]: Please fix these slashed to be "/" direction.

File: source/main.cpp [hunk 2]
```diff
@@ -3,31 +3,37 @@
...
-#include "source/NetworkInterface/ServerNetworkInterface.hpp"
-#include "source/PlayerAPI/Player.hpp"
-#include <iostream>
-#include <vector>
+#include<sstream>
+#include<iostream>
+#include<vector>
+#include"source\PlayerAPI\Player.hpp"
+#include"source\NetworkInterface\ServerNetworkInterface.hpp"
+#include"Lobby.hpp"
```
Comment (katiesweet) [inline source/main.cpp:11]: Same with these slashes

File: source/Lobby.hpp [hunk 3]
```diff
@@ -1,33 +1,82 @@
...
+#include <boost\archive\text_oarchive.hpp>
+#include <boost\archive\text_iarchive.hpp>
+#include <boost\algorithm\string.hpp>
+/// Application Includes
+#include "source\Messages\LobbyGame.hpp"
+#include "source\PlayerAPI\Player.hpp"
+#include "source\PlayerAPI\Game.hpp"
+#include "source\GameLogic\CrazyEightsLogic.hpp"
+#include "source\GameLogic\HeartsGame.hpp"
+#include "source\GameLogic\SpadesLogic.hpp"
```
Comment (katiesweet) [inline source/Lobby.hpp:22]: All these slashed need to be changed

File: source/NetworkInterface/mockClient.cpp [hunk 4]
```diff
@@ -1,6 +1,74 @@
...
+#include<string>
 
+#include <boost\serialization\access.hpp>
+#include <boost\archive\text_oarchive.hpp>
+#include <boost\archive\text_iarchive.hpp>
+#include <boost\algorithm\string.hpp>
+
+#include "ClientNetworkInterface.hpp"
+#include "source\Messages\LobbyGame.hpp"
+
```
Comment (katiesweet) [inline source/NetworkInterface/mockClient.cpp:11]: These as well.

**Review by katiesweet — CHANGES_REQUESTED** [review]
Will you please fix these slashed to be "/" way instead of "\" way @droidkfx? It won't build on Linux the way it is.

Also, will you please look at issue #64 ?

**Comment by droidkfx** [comment]
Everything looks ready, also fixed #64

**Review by droidkfx — COMMENTED** [review]
I think everything is ready to go. I fixed what was requested

**Comment by C-Merrill** [comment]
Nothing should be linked to the Messages library, so I fixed that. It isn't complete yet, but teams should be able to take copy the entire folder into their client project to include for deserialization of lobbygame archives.

**Comment by C-Merrill** [comment]
It was missing a slash in one include, but now it builds on Windows and all tests pass. @katiesweet can you verify build and tests on linux again?

**Comment by katiesweet** [comment]
I get this build error...

screen shot 2016-11-30 at 1 56 03 pm

@droidkfx @C-Merrill

**Comment by C-Merrill** [comment]
@katiesweet I did a minor fix, but it builds both ways on mine, so I have no idea if it worked? Can you try some things if it doesn't?

**Comment by droidkfx** [comment]
@C-Merrill @katiesweet I also get a clean build. but it seems to be complaining about the fact that the function returns a reference.

**Comment by mikamerath** [comment]
@droidkfx that's exactly what's happening, you can't create a reference to a non const temporary value

**Comment by droidkfx** [comment]
@michaelkamerath Well what should we do to fix it then. It needs to return a reference so that the lobby can edit the actual game being found, this is just an edge case handler that is causing the problem.

**Comment by mikamerath** [comment]
@droidkfx I would create a variable of GameType::UNKNOWN and pass that into the function

**Comment by katiesweet** [comment]
Now it builds, runs just fine, and all the tests pass, but it still throws a warning.

screen shot 2016-11-30 at 2 30 34 pm

I can look at it more if you need, but I won't be available to do so until about 4:00.

**Comment by droidkfx** [comment]
@michaelkamerath I don't understand what you mean or how that would fix the problem.

**Comment by C-Merrill** [comment]
It looks like the reference to a temporary object is dangerous, I guess. We could do what Michael's saying and have a lobbygame member of lobby that we can return a reference to, or we can make the function return a pointer and return a nullptr instead of an undefined game.

**Comment by mikamerath** [comment]
@droidkfx basically what @C-Merrill is saying, he just said it much better. I would prefer the pointer route but either one is a viable option.

**Comment by C-Merrill** [comment]
@katiesweet try this at least as a temporary solution. We may change to pointers, but this will work, I think

**Comment by mikamerath** [comment]
If it works and builds on both Windows and Linux and passes the unit tests after this, we'll merge it in. Something that I'll be doing over the next few days is trying to solve the many compiler warning we receive when building.

**Comment by droidkfx** [comment]
> If it works and builds on both Windows and Linux and passes the unit tests after this, we'll merge it in. 


@michaelkamerath @C-Merrill Yay, if it fits it ships!!

**Comment by katiesweet** [comment]
@michaelkamerath @C-Merrill  @droidkfx We are good to go! 👍

**Review by C-Merrill — APPROVED** [review]

**Pull request merged**